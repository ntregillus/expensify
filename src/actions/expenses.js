import uuid from 'uuid';
import database from '../firebase/firebase';
// react life cycle
/*
 * component calls action generator
 * action generator return object
 * component dispatches object
 * reduct store changes
 * 
 * new work with redux plugin
 * 
 * component calls action generator
 * action generator returns a FUNCTION (not object as above)
 * component dispatches function
 * function runs (which can trigger other actions!)
 * 
 */

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { //setting defaults if no properties are in expenseData
            description='', note='', amount=0,createdAt=0 
        } = expenseData;
        const newExpense = {description, note, amount, createdAt};
        //returning firebase promist to allow additional promis chaining
        return database.ref('expenses').push(newExpense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...newExpense
            }));
        });
    };
};

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense    
});
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});