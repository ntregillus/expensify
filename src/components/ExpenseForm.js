import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


const now = moment();
console.log(now.format('MMM Do YYYY'));

class ExpenseForm extends React.Component {
    state = {
        description: '',
        amount: '',
        note: '',
        createdAt: moment(),
        calFocused: false,
        error: ''
    };
    constructor(props){
        super(props);
        this.state = {
            description: '',
            amount: '',
            note: '',
            ...(props.expense||{createdAt: moment()}),
            calFocused: false,
            error: ''
        };
        if(props.expense)
        {
            this.state.createdAt = moment(props.expense.createdAt);
            this.state.amount = (props.expense.amount/100).toString();
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
    if(!amount ||amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}));
        }
    }
    onCalFocusChange = ({focused}) => this.setState((state)=> ({calFocused: focused}))
    createExpense = (e) =>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error:'Description and Amount are required'}));
        }else{
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description:this.state.description ,
                amount: parseFloat(this.state.amount)*100, //we need in cents, not dollars
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
             });
        }



    }
    render () {
        return (  
            <form onSubmit={this.createExpense}
                className="form">
                {   this.state.error 
                    && <p className="form__error" >{this.state.error}</p>
                }
                <input 
                    className="text-input"
                    type="text" 
                    value={this.state.description}  
                    onChange={this.onDescriptionChange}
                    placeholder="description" autoFocus />
                <input 
                    className="text-input"
                    type="text" 
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    placeholder="amount" />

                <SingleDatePicker 
                    date={this.state.createdAt} 
                    onDateChange={this.onDateChange}
                    focused={this.state.calFocused}
                    onFocusChange={this.onCalFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(day)=> false}
                    />
                <textarea 
                    className="textarea"
                    placeholder="Add a note for expense"
                    value={this.state.note}
                    onChange={this.onNoteChange}>
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
}
export default ExpenseForm;