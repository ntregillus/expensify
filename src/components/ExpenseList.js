import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import ExpensesTotal from './ExpensesTotal';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
    {
        props.expenses.length  === 0 ? (
            <p>No Expenses</p>
          )  : (
            props.expenses.map( (expense) => {
                return (
                    <div key={expense.id} >
                        <ExpenseListItem {...expense} /> 
                        <hr />
                    </div>
                );
            })
        )
    }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
    };
};

export default connect(mapStateToProps)(ExpenseList);
