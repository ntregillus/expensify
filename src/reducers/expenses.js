
const expensesReducerDefaultState = [];
const expensesReducer = (state=expensesReducerDefaultState, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            const result = [...state, action.expense];
            return result;
        case 'REMOVE_EXPENSE':
        const r = state.filter((item)=> item.id != action.id);
        return r;
        case 'EDIT_EXPENSE':
        const newResult = state.map((expense)=>{
            if(expense.id === action.id){
                return {
                    ...expense, 
                    ...action.updates
                };
            }
            else{
                return expense;
            }
        });
        return newResult;
        default:
        return state;
    }
};

export default expensesReducer;