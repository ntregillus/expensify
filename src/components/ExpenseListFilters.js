import React from 'react';
import {DateRangePicker} from 'react-dates';

import {connect} from 'react-redux';
import { setTextFilter } from '../actions/filters';
import { sortByAmount, sortByDate,setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component{
    state = {
        calFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate({startDate});
        this.props.setEndDate({endDate});
    }
    onFocusChange = (calFocused) => {
        this.setState(()=> ({calFocused}));
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    onSortChange = (e) => {
        if(e.target.value === 'date'){                
            this.props.sortByDate();
        }else{
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    focusedInput={this.state.calFocused}
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=> false}
                    showClearDates={true}
                />
            </div>
        );
    }

} 


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};
const mapDispatchToProps = (dispatch) => ({
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);