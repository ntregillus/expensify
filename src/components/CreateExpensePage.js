import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect } from 'react-redux'
import PageHeader from './PageHeader';
import {startAddExpense} from '../actions/expenses';

export class CreateExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <PageHeader title="Create Expense" />
                <div className="content-container">
                    <ExpenseForm onSubmit={this.onSubmit}/>
                </div>           
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(startAddExpense(expense))
});
export default connect(undefined, mapDispatchToProps)(CreateExpensePage);