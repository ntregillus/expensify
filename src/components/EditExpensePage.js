import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import PageHeader from './PageHeader';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.match.params.id, expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.removeExpense({id: this.props.match.params.id});
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <PageHeader title="Edit Expense" />
                <div className="content-container">
                    <ExpenseForm 
                        onSubmit={this.onSubmit} 
                        expense={this.props.expense}    
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>
                        Remove Expense
                    </button>
                </div>
            </div>
        );
    }
}


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(({id})=> props.match.params.id == id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    removeExpense: (expense) => dispatch(startRemoveExpense(expense))
});
    
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);