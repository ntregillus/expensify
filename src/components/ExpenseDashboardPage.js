import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesTotal from './ExpensesTotal';
export const ExpenseDashboardPage = () => (
    <div>
        <ExpensesTotal />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);


export default ExpenseDashboardPage;