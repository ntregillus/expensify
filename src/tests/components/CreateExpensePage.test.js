import React from 'react';
import {shallow} from 'enzyme';
import {CreateExpensePage} from '../../components/CreateExpensePage';
import expenses from '../fixtures/expenses';


let onSubmitSpy, history, wrapper;
beforeEach(() => {
    onSubmitSpy = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(
        <CreateExpensePage addExpense={onSubmitSpy} history={history}
    />);
});

test('should render CreateExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () =>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[1]);
});