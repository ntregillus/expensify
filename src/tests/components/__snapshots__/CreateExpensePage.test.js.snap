// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreateExpensePage correctly 1`] = `
<div>
  <h1>
    Expense Page
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
